R/plotly.R

```diff
@@ -254,48 +254,64 @@ plotly_build <- function(l = last_plot()) {
       # start processing specially named arguments
-      s <- dat[["size"]]
+      s <- trace[["size"]]
       if (!is.null(s)) {
         if (!is.numeric(s)) warning("size should be numeric", call. = FALSE)
         # if autosizing is used, guess that the plot is 300 by 600
-        auto <- dat[["layout"]][["autosize"]] %||% TRUE
+        auto <- trace[["layout"]][["autosize"]] %||% TRUE
         hw <- if (auto) c(300, 600)
-        else c(dat[["layout"]][["height"]], dat[["layout"]][["width"]])
+        else c(trace[["layout"]][["height"]], trace[["layout"]][["width"]])
         # ensure that markers cover 30% of the plot area
         m <- list(
           size = 0.3 * prod(hw) * (s/sum(s)),
           sizemode = "area"
         )
         # the marker object is the only type of object which respects size
-        dat[["marker"]] <- modifyList(dat[["marker"]] %||% list(), m)
+        trace[["marker"]] <- modifyList(trace[["marker"]] %||% list(), m)
         # either add some appropriate hover text
         txt <- paste0(as.list(d$args)[["size"]], " (size): ", s)
-        dat[["text"]] <- if (is.null(dat[["text"]])) txt else paste0(dat[["text"]], "<br>", txt)
+        trace[["text"]] <- if (is.null(trace[["text"]])) txt else paste0(trace[["text"]], "<br>", txt)
       }
-      has_color <- !is.null(dat[["color"]]) || 
-        isTRUE(!is.null(dat[["z"]]) && !dat[["type"]] %in% "scatter3d")
-      has_symbol <- !is.null(dat[["symbol"]])
-      has_group <- !is.null(dat[["group"]])
-      if (has_color) {
+      # put the whole dat into a single trace first
+      traces <- list(trace)
+      if (!is.null(trace[["color"]]) || 
+          isTRUE(!is.null(trace[["z"]]) && !trace[["type"]] %in% "scatter3d")) {
         title <- as.list(d$args)[["color"]] %||% as.list(d$args)[["z"]] %||% ""
-        x$data <- c(x$data, colorize(dat, title))
+        traces <- colorize(traces, dat, title)
       }
       # TODO: add a legend title (is this only possible via annotations?!?)
-      if (has_symbol) x$data <- c(x$data, symbolize(dat))
-      if (has_group) x$data <- c(x$data, traceify(dat, "group"))
-      if (!has_color && !has_symbol && !has_group) x$data <- c(x$data, list(dat))
+      if (!is.null(trace[["symbol"]])) traces <- symbolize(traces, dat)
+      if (!is.null(trace[["group"]])) traces <- subdivide_traces(traces, dat, "group")
+      if (!is.null(trace[["legenditem"]])) {
+        legenditems <- unique(trace[["legenditem"]])
+        showlegend <- rep.int(TRUE, length(legenditems))
+        traces <- subdivide_traces(traces, dat, "legenditem", function(sub_trace, lvl) {
+          sub_trace$legendgroup <- legenditems[[lvl]]
+          sub_trace$name <- legenditems[[lvl]]     # override the trace name by legendgroup
+          sub_trace$showlegend <- showlegend[[lvl]] # put each group to legend once
+          if (showlegend[[lvl]]) {
+            showlegend[lvl] <<- FALSE
+          }
+          sub_trace
+        })
+      }
```

Comment:
@cpsievert It tries to solve the problem of having an item in the legend for each trace (which is not what you want if e.g. you are doing something similar to `geom_segment()`). `legenditem` specifies the name of the _single_ element in the legend that all the traces in the same `legenditem`-group would be referring to (unless they are further subdivided by something else).
To implement this, only the first trace of the group are shown in the legend, but all the other traces still share the same `legendgroup`.
Although it utilizes `legendgroup` mechanism, I thought it's better to call the property `legenditem`, because from the user POV it doesn't do any grouping of legend elements, it just creates one item for the group of traces.